group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -898,31 +987,44 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             group.setMetadataRefreshDeadline(currentTimeMs + consumerGroupMetadataRefreshIntervalMs, groupEpoch);
         }
 
-        // 2. Update the target assignment if the group epoch is larger than the target assignment epoch. The
-        // delta between the existing and the new target assignment is persisted to the partition.
+        // 2. Update the target assignment if the group epoch is larger than the target assignment epoch or a static member
+        // replaces an existing static member. The delta between the existing and the new target assignment is persisted to the partition.
         int targetAssignmentEpoch = group.assignmentEpoch();
         Assignment targetAssignment = group.targetAssignment(memberId);
-        if (groupEpoch > targetAssignmentEpoch) {
+        if (groupEpoch > targetAssignmentEpoch || staticMemberReplaced) {
             String preferredServerAssignor = group.computePreferredServerAssignor(
                 member,
                 updatedMember
             ).orElse(defaultAssignor.name());
 
             try {
-                TargetAssignmentBuilder.TargetAssignmentResult assignmentResult =
-                    new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor))
+                TargetAssignmentBuilder assignmentResultBuilder =
+                    new TargetAssignmentBuilder(groupId, groupEpoch, assignors.get(preferredServerAssignor));
+                TargetAssignmentBuilder.TargetAssignmentResult assignmentResult;
+                // A new static member is replacing an older one with the same subscriptions.
+                // We just need to remove the older member and add the newer one. The new member can
+                // reuse the target assignment of the older member.
+                if (staticMemberReplaced && groupEpoch == targetAssignmentEpoch) {
+                    targetAssignment = group.targetAssignment(existingStaticMember.memberId());
+                    assignmentResult = assignmentResultBuilder
+                        .removeMember(existingStaticMember.memberId())
+                        .addOrUpdateMember(memberId, updatedMember)
+                        .build();
+                    records.addAll(assignmentResult.records());
+                } else {
+                    assignmentResult = assignmentResultBuilder
                         .withMembers(group.members())
                         .withSubscriptionMetadata(subscriptionMetadata)
                         .withTargetAssignment(group.targetAssignment())
                         .addOrUpdateMember(memberId, updatedMember)
```

Comment:
When the `TargetAssignmentBuilder` builds the spec for the assignor, it must use the target assignment of the previous static member for the new static member. How do we ensure this? We may have to update the `TargetAssignmentBuilder` to understand that a static member is replaced.

Reply:
I think I got it now. I added some state tracking to the `TargetAssignmentBuilder` so that it doesn't compute assignments for a replacing static member.